### PR TITLE
Trim per-allele FORMAT annotations and optionally retain raw AS annotations

### LINKED ISSUE #5704: F1R2 and F2R2 annotations not updated by GenotypeGvcfs
The gVCF output of `HaplotypeCaller`:
```
... ATTTTTT A,ATT,ATTT,ATTTT,ATTTTT,<NON_REF> ... GT:AD:DP:F1R2:F2R1:GQ:PL:SB	4/5:8,2,5,8,19,17,0:59:5,2,5,8,17,17,0:3,0,0,0,2,0,0:99:841,879,2162,594,1484,1406,395,983,921,881,140,419,309,264,313,382,470,281,174,0,402,828,1443,1223,962,474,507,1363:5,3,49,2
```

The VCF output of `GenotypeGvcfs`:
```
... ATT A,AT ... GT:AD:DP:F1R2:F2R1:GQ:PL ... 1/2:8,19,17:59:5,2,5,8,17,17,0:3,0,0,0,2,0,0:99:841,140,313,382,0,402
```

In both cases, `F1R2` is `5,2,5,8,17,17,0`.   There are seven alleles in the gVCF but only three in the VCF, so the `F1R2` is wrong for the latter.

### PR DESCRIPTION
**Pull request by ldgauthier**
GenotypeGVCFs now uses the header info to determine if FORMAT lists need to be subset when alleles are dropped. Fixes #5704

Added an option to keep raw AS values like AS_SB_TABLE and AS_QUAL, as requested in #5698 (does not close that issue)

### CODE REVIEW

**Comment by ldgauthier** [comment]
@nh3 care to take a look?  In the GGVCFs integration test, the with OxoGReadCounts file has 
```
20      10101674        .       TTGTGTG T,TTG,TTGTGTGTGTGTG,TTGTGTGTGTGTGTG,<NON_REF>   1464.10 .       DP=64;ExcessHet=3.0103;MLEAC=0,1,1,0,0;MLEAF=0.00,0.500,0.500,0.00,0.00;RAW_MQandDP=196622,64   GT:AD:DP:F1R2:F2R1:GQ:PL:SB         2/3:0,3,22,6,4,0:35:0,1,13,3,3,0:0,2,9,3,1,0:47:1481,1115,1205,557,577,581,932,574,0,925,973,621,47,860,972,1495,1189,615,968,1015,1570:0,0,20,15
```
with read counts

| TTGTGTG* | T | TTG | TTGTGTGTGTGTG | TTGTGTGTGTGTGTG | <NON_REF> |
| --- | -- | -- | -- | -- | -- |
| 0 | 1 | 13 | 3 | 3 | 0 |
| 0 | 2 | 9 | 3 | 1 | 0 |

after genotyping alleles get dropped and trimmed and we have

| TTGTG* |  T | TTGTGTGTGTG |
| -- | -- | -- |
| 0 | 13 | 3 |
| 0 | 9 | 3 |

Do you agree that's as expected?  (I couldn't reproduce your exact example without the exact bam.)

**Comment by ldgauthier** [comment]
Also @tfenne hopefully this meets expectations for AS_SB_TABLE (which I have now added in AS_Annotations.keepRawCombined.expected.vcf)

**Comment by nh3** [comment]
I believe this was meant for @nh13, not me.

**Comment by ldgauthier** [comment]
Sorry.  I don't think that's the only one.  Bad case of the Mondays.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 1]
```diff
@@ -395,6 +408,29 @@ private VariantContext removeNonRefAlleles(final VariantContext vc) {
         }
     }
 
+    private GenotypesContext subsetAlleleSpecificFormatFields(final VCFHeader outputHeader, final GenotypesContext originalGs, final int[] relevantIndices) {
+        final GenotypesContext newGTs = GenotypesContext.create(originalGs.size());
+        GenotypeBuilder gb;
```
[the remark this answers is not in the document]
Comment (ldgauthier) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: I seem to remember someone telling me that it's more efficient to reuse the same VariantContextBuilder than declare more, but maybe that's less of an issue with GenotypeBuilders or not true at all.

[the remark this answers is not in the document]
Comment (nh13) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Your overwriting the reference, so not re-using it, I think.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 2]
```diff
@@ -127,6 +128,12 @@
     @Argument(fullName=CombineGVCFs.ALLELE_FRACTION_DELTA_LONG_NAME, doc = "Margin of error in allele fraction to consider a somatic variant homoplasmic")
     protected double afTolerance = 1e-3;  //based on Q30 as a "good" base quality score
 
+    /**
+     * If specified, keep the combined raw annotations (e.g. AS_SB_TABLE) after genotyping.  This is applicable to Allele-Specific annotations
+     */
+    @Argument(fullName=KEEP_COMBINED_LONG_NAME, shortName = KEEP_COMBINED_SHORT_NAME, doc = "If specified, keep the combined raw annotations")
+    protected boolean keepCombined = false;
```
Comment (tfenne) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: Thanks @ldgauthier!  This looks great.

**Comment by tfenne** [comment]
Thanks again for getting this done and merged @ldgauthier & @droazen!

**Pull request merged**